File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    // Function to detect cycle in a directed graph.
    
    bool isCycle(int u,vector<int>adj[],vector<int>&vis,vector<int>&cvis){
        vis[u] = 1;
        cvis[u] = 1;
        for(int v:adj[u]){
            if(vis[v]){
                if(cvis[v]){
                    return true;
                }
            }
            else{
                bool flag = isCycle(v,adj,vis,cvis);
                if(flag) return true;
            }
        }
        cvis[u] = 0;
        return false;
    }
    
    bool isCyclic(int n, vector<int> adj[]) {
        vector<int>visited(n,0);
        vector<int>cvisited(n,0);
        for(int i=0;i<n;i++){
            if(visited[i]) continue;
            if(isCycle(i,adj,visited,cvisited)){
                return true;
            }
        }
        return false;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
        }

        Solution obj;
        cout << obj.isCyclic(V, adj) << "\n";
    }

    return 0;
}

// } Driver Code Ends

File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class DriverClass {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            int V = sc.nextInt();
            int E = sc.nextInt();
            for (int i = 0; i < V; i++)
                list.add(i, new ArrayList<Integer>());
            for (int i = 0; i < E; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                list.get(u).add(v);
            }
            if (new Solution().isCyclic(V, list) == true)
                System.out.println("1");
            else
                System.out.println("0");
        }
    }
}
// } Driver Code Ends


/*Complete the function below*/

class Solution {
    // Function to detect cycle in a directed graph.
    
    public boolean isCycle(int u,ArrayList<ArrayList<Integer>>adj,boolean vis[],boolean cvis[]){
        vis[u] = true;
        cvis[u] = true;
        for(int v:adj.get(u)){
            if(vis[v] == true){
                if(cvis[v] == true){
                    return true;
                }
            }
            else{
                boolean flag = isCycle(v,adj,vis,cvis);
                if(flag) return true;
            }
        }
        cvis[u] = false;
        return false;
    }
    
    public boolean isCyclic(int n, ArrayList<ArrayList<Integer>> adj) {
        boolean visited[] = new boolean[n];
        boolean cvisited[] = new boolean[n];
        for(int i=0;i<n;i++){
            if(visited[i] == true) continue;
            if(isCycle(i,adj,visited,cvisited) == true){
                return true;
            }
        }
        return false;
    }
}